### PR TITLE
New Windows Persistence module

### PR DESCRIPTION
**Pull request by drforbin**
This is persistence module I use in field.
It's very useful for uploading a payload created with veil etc.
One of the problems with the current persistence module is that AV sees it,
this gets around that. Use it with my other module winprivs.
Give me your feedback thanxs!!
## Verification

List the steps needed to make sure this thing works
- [ ] Start `msfconsole`
- [ ] `use post/windows/manage/persistence_exe`
- [ ] set session to desired
- [ ] **Verify** the thing does what it should
- [ ] **Verify** the thing does not do what it should not

### CODE REVIEW

**Comment by drforbin** [comment]
Please let me know any ideas to improve.

Thank you.

**Comment by Megaeloelo** [comment]
How it works with Shellter?

**Comment by drforbin** [comment]
I mainly try and stay clear of closed source projects BUT that being said.
Any independent executable should work fine. I've used this and it works.
Alot of times I design my own payloads and this can be used to upload (persist) them as well.
give it a try!

**Comment by mubix** [comment]
have you tried using the EXE::Custom option in the standard persistence module?

**Comment by drforbin** [comment]
No....I use to use the older persistence module. In fact what happened is I added the EXE feature to Carlos's module sometime ago and I noticed it was changed so I thought I would place this module back online to see if it may be of benefit.

**Comment by mubix** [comment]
check out `exploit/windows/local/persistence`, it allows you to easily specify another EXE with the `set EXE::Custom /root/myveildbinary.exe`

**Comment by drforbin** [comment]
Thats all well and good. But I think it still wraps it in the VBS stuff.

**Comment by wchen-r7** [comment]
I see that all the verification steps are checked, but who's doing the verification?? I don't see an assignee.

**Comment by bcook-r7** [comment]
I'm unchecking them.

**Comment by wchen-r7** [comment]
@drforbin I'm looking at the code, and your #write_exe_to_target method does not make a lot of sense to me. The first argument is named ```vbs```, but it doesn't look like you're doing any VBS code in it. It looks like all it's doing is loading ```datastore['REXE']``` to the target machine. It doesn't look like the naming correctly describes what the code is doing?

Also, datastore['REXE'] probably should be OptPath.

However, the biggest reason that's holding me back from landing is this seems to be rather similar to exploit/windows/local/persistence. And looking at the conversation, the only reason you wanted to create this module in the first place is because exploit/windows/local/persistence's VBS code is being flagged by AV? If that's the only reason, you know we can just update exploit/windows/local/persistence, and make VBS optional.

**Comment by wchen-r7** [comment]
Marking this as delayed because I feel like there's a debate between just improving exploit/windows/local/persistence, or add a new one that seems to be slightly different.

**Comment by drforbin** [comment]
Hi..I checked the verification. I thought that is what your suppose to do, sorry!
I have not used this system before.
I did write winprivs a long while ago which is currently in the repo.
Sorry for misunderstanding.

As for improving carlo's work I tried that a while back and it was reverted.
So I thought I'd just submit this module which I hacked together in field.
I will clean up the target function tonight.

It is useful.

**Comment by drforbin** [comment]
I cleaned up target method and changed option variables.
Please see commit.

**Comment by void-in** [comment]
@drforbin The REXE type should be OptPath as suggested by @wchen-r7.
```ruby
OptPath('REXE', [true, 'The remote executable to use']),
```
You should remove the default value as well as shown above. The advantage of OptPath instead of OptString is that the datastore option is doing to be validated before the module execution.

File: modules/post/windows/manage/persistence_exe.rb [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+        host,port = session.tunnel_peer.split(':')
+        @clean_up_rc = ""
+
+
+        if datastore['REXEPATH'] == nil
+            print_error ("REXE is null...please define")
+            return
+        end
+
+        if not ::File.exist?(datastore['REXEPATH'])
```
Comment (nonmadden) [inline modules/post/windows/manage/persistence_exe.rb]: `unless ::File.exist?(datastore['REXEPATH'])`

File: modules/post/windows/manage/persistence_exe.rb [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+        print_status("Running module against #{sysinfo['Computer']}")
+
+        # Set vars
+        rexe = datastore['REXEPATH']
+        rexename = datastore['REXENAME']
+        host,port = session.tunnel_peer.split(':')
+        @clean_up_rc = ""
+
+
+        if datastore['REXEPATH'] == nil
```
Comment (nonmadden) [inline modules/post/windows/manage/persistence_exe.rb]: `if datastore['REXEPATH'].nil?`

**Comment by nonmadden** [comment]
can you writing code ruby style?
ref: https://github.com/bbatsov/ruby-style-guide#syntax

File: modules/post/windows/manage/persistence_exe.rb [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+            end
+
+            # Create the log directory
+            ::FileUtils.mkdir_p(logs)
+
+            #logfile name
+            logfile = logs + ::File::Separator + Rex::FileUtils.clean_path(host + filenameinfo) + ".rc"
+            return logfile
+
+end
```
Comment (nonmadden) [inline modules/post/windows/manage/persistence_exe.rb]: indent

**Comment by drforbin** [comment]
ok..thanks guys...will clean up tonight.

**Comment by drforbin** [comment]
I changed formatting and am now using OptPath.
With the use of OptPath I am assuming that it validates both that OptPath has been defined and that file is present.  If this not the behavior let me know and I will fix it.

File: modules/post/windows/manage/persistence_exe.rb [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+        #Get hostname
+        host = session.sys.config.sysinfo["Computer"]
+
+        # Create Filename info to be appended to downloaded files
+        filenameinfo = "_" + ::Time.now.strftime("%Y%m%d.%M%S")
+
+        # Create a directory for the logs
+        if log_path
+            logs = ::File.join(log_path, 'logs', 'persistence', Rex::FileUtils.clean_path(host + filenameinfo) )
+            else
```
Comment (nonmadden) [inline modules/post/windows/manage/persistence_exe.rb]: wrong indent?

File: modules/post/windows/manage/persistence_exe.rb [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+    # Function to install payload in to the registry HKLM or HKCU
+    #-------------------------------------------------------------------------------
+    def write_to_reg(key,script_on_target)
+
+        nam = Rex::Text.rand_text_alpha(rand(8)+8)
+        print_status("Installing into autorun as #{key}\\Software\\Microsoft\\Windows\\CurrentVersion\\Run\\#{nam}")
+        if(key)
+            registry_setvaldata("#{key}\\Software\\Microsoft\\Windows\\CurrentVersion\\Run",nam,script_on_target,"REG_SZ")
+            print_good("Installed into autorun as #{key}\\Software\\Microsoft\\Windows\\CurrentVersion\\Run\\#{nam}")
+            else
```
Comment (nonmadden) [inline modules/post/windows/manage/persistence_exe.rb]: wrong indent

File: modules/post/windows/manage/persistence_exe.rb [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+    #-------------------------------------------------------------------------------
+    def install_as_service(script_on_target)
+
+        if  is_system? or is_admin?
+            print_status("Installing as service..")
+            nam = Rex::Text.rand_text_alpha(rand(8)+8)
+            print_status("Creating service #{nam}")
+            service_create(nam, nam, "cmd /c \"#{script_on_target}\"")
+            @clean_up_rc << "execute -H -f sc -a \"delete #{nam}\"\n"
+            else
```
Comment (nonmadden) [inline modules/post/windows/manage/persistence_exe.rb]: wrong intent

**Comment by drforbin** [comment]
reformatted else statements to conform with proper ruby indent rules.

**Comment by join-us** [comment]
Maybe you can use ```rubocop``` and ```msftidy``` to check the code style.

**Comment by drforbin** [comment]
It's a straight remote installer. You create the executable it will install it.
The problem with the other is the wrapper (VBS) etc. This is a stripped down version of Carlos's  persistence.rb. It works very well. As I stated I've used this in field, where I could not get the other to work.

**Comment by Megaeloelo** [comment]
For example, what if the payload is in the calculator?
You will see the calculator every time you start the system?

**Comment by drforbin** [comment]
Pretty much, yes.

**Comment by drforbin** [comment]
Oh..this module is also very useful for planting keyloggers etc.

**Comment by busterb** [comment]
removing the delayed flag to reevaluate

**Comment by dmaloney-r7** [comment]
# Release Notes

Adds drforbin's new EXE persistence module for creating persistent sessions on a compromised machine

**Comment by drforbin** [comment]
Thank you!

**Pull request merged**